### ISSUE TITLE
Section on Connecting to a swarm service from External Clients

### ISSUE THREAD

**Issue by kolotyluk** [docker/docker.github.io#4406]
File: [docker-for-aws/index.md](https://docs.docker.com/docker-for-aws/), CC @johndmulhausen 

After following the instructions here to the best of my ability, I cannot connect to my services in the swarm from external clients.

1. For some reason it is not possible to connect to worker nodes from external clients, even after configuring the 'Node Security Group' the same as the 'Swarm wide access' security group.

2. When you us `docker stack deploy . . .` it seems to deploy services randomly across manager and worker nodes.

3. The ELB is configured to route to all nodes in the cluster, but the worker nodes cannot be configured to accept inbound connections for some reason.

4. It is not clear how, or if it is even possible, to define a permanent public address to connect to services in the swarm, when `docker stack` just randomly moves things around.

**Comment by mstanleyjones** [comment]
@friism who would be able to help with this?

**Comment by friism** [comment]
@FrenchBen

**Comment by friism** [comment]
> After following the instructions here to the best of my ability, I cannot connect to my services in the swarm from external clients.

Please provide a `docker-compose.yml` file that reproduces the problem (or `docker service create` command if you prefer).

When you've deployed a service, it's available on the default dns target (you can see this in the cfn outputs): https://docs.docker.com/docker-for-aws/deploy/#run-apps

> For some reason it is not possible to connect to worker nodes from external clients, even after configuring the 'Node Security Group' the same as the 'Swarm wide access' security group.

Why do you want to do that? We recommend connecting to deployed services using the load balancer that Docker for AWS configures and auto-updates for you.

> The ELB is configured to route to all nodes in the cluster, but the worker nodes cannot be configured to accept inbound connections for some reason.

They'll accept connections from the ELB - do you have evidence to the contrary? (other than not being able to connect to them from the internet at large, which is by design).

> It is not clear how, or if it is even possible, to define a permanent public address to connect to services in the swarm, when docker stack just randomly moves things around.

It's the ELB backing the default DNS target.

**Comment by kolotyluk** [comment]
Will respond more when I understand your response.

'cfn outputs' does not mean anything to me, nor it is mentioned in https://docs.docker.com/docker-for-aws/deploy

**Comment by FrenchBen** [comment]
@kolotyluk We've detailed here how to connect to worker nodes:
https://docs.docker.com/docker-for-aws/deploy/#worker-nodes

Docker stack deploy deploys services that correspond to your YML file declaration - For more information on how services are deployed, please see:
https://docs.docker.com/engine/swarm/services/

As you noticed, the ELB has been defined to target all nodes in the cluster, which means that you can always use the ELB as your constant endpoint to access the services you've deployed to your swarm.

**Comment by kolotyluk** [comment]
Here is where I am now.

1. I can ping the ELB and get a response.

2. I cannot telnet to my Kafka brokers via the ELB. I get back `Connection refused`

3. I can telnet to my Kafka brokers via the AWS instance public name/address. This indicates the brokers running in the swarm are reachable, as I can see in the logs the incoming telnet connections. I have them running on manager nodes for now as dealing with ELB issues is just a distraction I don't need at the moment.

4. I cannot connect to my brokers via the Kafka Admin Client via the public name/address not sure why.

5. I am using `docker stack . . .` with a yml file to bring up my services. Unfortunately the Kafka Brokers keep restarting with `task: non-zero exit (137)`. There is nothing in the Kafka Broker logs to suggest there is anything wrong with them. I am wondering if Docker is restarting them for some reason (some articles I found on the internet suggest this may be the case).

6. https://docs.docker.com/engine/swarm/services is a very painful way to manage services, which is why I am trying to use `docker stack` but some people have mentioned this is still experimental, so I may have to resort to `docker service . . .`

7. https://docs.docker.com/docker-for-aws/deploy/#worker-nodes is a non-issue at the moment because I can configure `docker stack . . .` to deploy on manager nodes for testing.

8. All in all I chose to try using Docker Swarm CloudFormation to try to get my Kafka service running on AWS, but it has been over a week of *intense* frustration. I have Kafka running fine in our developer environment using `docker-compose`, but the transition to AWS is astonishingly more difficult. It is nice to have CloudFormation create the AWS Stack, but some of the documentation on how to use it could be better.

Basically what is most frustrating of all is the lack of good information on troubleshooting. Your documentation is not alone in this, as diagnostics is generally ignored by most organizations. Similarly, people try to offer good quick-start documentation, but it is generally limited to hello-world and of limited use.
